org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/EitherTypeAdapter.java

```diff
@@ -201,16 +203,30 @@ public void write(JsonWriter out, Either<L, R> value) throws IOException {
 	protected Either<L, R> createLeft(L obj) throws IOException {
 		if (Either3.class.isAssignableFrom(typeToken.getRawType()))
 			return (Either<L, R>) Either3.forLeft3(obj);
-		else
-			return Either.forLeft(obj);
+		else if ("org.eclipse.lsp4j.ProgressToken".equals(typeToken.getRawType().getName())) {
```

Comment:
This looks a little spooky to me. If you follow the original suggestion by @spoenemann and remove the type ProgressToken, this hidden inverse dependency would become obsolete. Why do we need the ProgressToken type at all?

Reply:
> Why do we need the ProgressToken type at all?

I proposed that because the type is used very often in the protocol. It's just a type alias (which does not exist in Java) that makes the protocol implementation more readable.

We can remove the class name check and always try to invoke the public constructor if it exists.

Reply:
In my opinion i prefere something more simple:
- no ProgressToken and replace the Either<String, Number>, like i do before
- If you absolutely want this class, i propose to add an adapter, this way the code is more simple and less hugly.

Reply:
Ok then revert back to `Either<String, Number>` as before.